dev-tools/mage/manifest/manifestspecs.go

```diff
@@ -0,0 +1,163 @@
+	BuildId              string             `json:"build_id"`
+	StartTime            string             `json:"start_time"`
+	EndTime              string             `json:"end_time"`
+	BuildDurationSeconds int                `json:"build_duration_seconds"`
+	ManifestVersion      string             `json:"manifest_version"`
+	Prefix               string             `json:"prefix"`
+	Projects             map[string]Project `json:"projects"`
+}
+
+type PackageComponent struct {
+	PackageName    string `yaml:"package_name"`
+	PackageRepo    string `default:"" yaml:"package_repo"`
+	PackageVersion string `yaml:"package_version"`
+	PackageBuildId string `default:"" yaml:"package_build_id"`
+}
+
+func downloadFile(url string, filepath string) (string, error) {
+	outFile, fileErr := os.Create(filepath)
+	if fileErr != nil {
+		return "", fmt.Errorf("failed to create destination file %w", fileErr)
+	}
+	defer func() {
+		outFile.Close()
+	}()
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return filepath, fmt.Errorf("failse to download manifest [%s]\n %w\n", url, err)
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(resp.Body)
+
+	_, errCopy := io.Copy(outFile, resp.Body)
+	if errCopy != nil {
+		return "", fmt.Errorf("failse to decode manifest response [%s]\n %w\n", url, err)
+	}
+	if mg.Verbose() {
+		log.Printf("<<<<<<<<< Downloaded: %s to %s", url, filepath)
+	}
+
+	return outFile.Name(), nil
+}
+
+func downloadFileWithRetries(url string, filepath string) (string, error) {
+	var err error
+	var resp string
+	for _, backoff := range backoffSchedule {
+		resp, err = downloadFile(url, filepath)
+		if err == nil {
+			break
+		}
+		if mg.Verbose() {
+			log.Printf("Request error: %+v\n", err)
+			log.Printf("Retrying in %v\n", backoff)
+		}
+		time.Sleep(backoff)
+	}
+
+	// All retries failed
+	if err != nil {
+		return resp, err
+	}
+	return resp, nil
+}
+
+func downloadManifestData(url string) (ManifestResponse, error) {
+	var response ManifestResponse
+	resp, err := http.Get(url)
+	if err != nil {
+		return response, fmt.Errorf("failse to download manifest [%s]\n %w\n", url, err)
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(resp.Body)
+
+	err = json.NewDecoder(resp.Body).Decode(&response)
+	if err != nil {
+		return response, fmt.Errorf("failse to decode manifest response [%s]\n %w\n", url, err)
+	}
+	return response, nil
+}
+
+func downloadManifestDataWithRetries(url string) (ManifestResponse, error) {
+	var err error
+	var resp ManifestResponse
+
+	for _, backoff := range backoffSchedule {
+		resp, err = downloadManifestData(url)
+
+		if err == nil {
+			break
+		}
+		if mg.Verbose() {
+			log.Printf("Request error: %+v\n", err)
+			log.Printf("Retrying in %v\n", backoff)
+		}
+		time.Sleep(backoff)
+	}
+
+	// All retries failed
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, nil
+}
```

Comment:
This function (and the one above could be the same using generics)
```suggestion
func DoWithRetries[T any](f func() (T, error)) (T, error) {
	var err error
	var resp T

	for _, backoff := range backoffSchedule {
		resp, err = f()

		if err == nil {
			break
		}
		if mg.Verbose() {
			log.Printf("Request error: %+v\n", err)
			log.Printf("Retrying in %v\n", backoff)
		}
		time.Sleep(backoff)
	}

	// All retries failed
	if err != nil {
		return resp, err
	}

	return resp, nil
}
```

Reply:
Done in 5450fef1b7420013955ffb989cc2392de62233f8